### PR TITLE
Purpose api change

### PR DESCRIPTION
**Pull request by veramine**
⚠️ DO NOT MERGE YET ⚠️
To be discussed in the Oct 18th '23 cohort meeting

This PR changes protobom to use an enum named Purpose for the PrimaryPurpose field that had been a string.  See issue https://github.com/bom-squad/protobom/issues/124 for more details on the problem this PR fixes.

However, it does not include support for types ```data```, ```device-driver```, ```machine-learning-model```, or ```platform``` because those fields were not in the original comment related to Purpose.  They would be easy to add, of course.  @puerco can you help set a final list of purposes?

This is the list of updates this PR includes:

- Update protobuf
- Regenerate go types from proto
- Update node and its test
- Update serializers
- Update unserializers
- Update new alpha SPDX3 code
- Update required protoc version to v4.24.4
- Regenerated proto in conformance test
- Update README to reflect API change

Will close https://github.com/bom-squad/protobom/issues/124 when approved.

### CODE REVIEW

**Comment by veramine** [comment]
Sorry that took a couple of tries to get right!  

@puerco: will need your review on this one.  Let me know if I missed anything!  Thanks!

File: api/sbom.proto [hunk 1]
```diff
@@ -53,6 +55,7 @@ message Node {
     repeated string file_types = 27; // File types
     map<int32,string> identifiers = 28;  // Software identifiers
     map<int32,string> hashes = 29;
+    Purpose primary_purpose = 30;
```
Comment (puerco) [inline api/sbom.proto:58]: In SPDX, the primary has a cardinality of one or more, so this should be a list:

```suggestion
    repeated Purpose primary_purpose = 30;
```

This obviously triggers changes in the dependent objects, sorry :/

File: pkg/native/serializers/serializer_cdx_test.go [hunk 2]
```diff
@@ -17,20 +17,20 @@ func TestComponentType(t *testing.T) {
...
+		"node package, pp file": {func(n *sbom.Node) { n.PrimaryPurpose = sbom.Purpose_FILE; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeFile},
+		"firmware":              {func(n *sbom.Node) { n.PrimaryPurpose = sbom.Purpose_FIRMWARE; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeFirmware},
+		"framework":             {func(n *sbom.Node) { n.PrimaryPurpose = sbom.Purpose_FRAMEWORK; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeFramework},
+		"operating-system":      {func(n *sbom.Node) { n.PrimaryPurpose = sbom.Purpose_OPERATINGSYSTEM; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeOS},
+		// "data":                  {func(n *sbom.Node) { n.PrimaryPurpose = "data"; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeData},
+		// "device-driver":         {func(n *sbom.Node) { n.PrimaryPurpose = "device-driver"; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeDeviceDriver},
+		// "machine-learning-model": {func(n *sbom.Node) { n.PrimaryPurpose = "machine-learning-model"; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypeMachineLearningModel},
+		// "platform": {func(n *sbom.Node) { n.PrimaryPurpose = "platform"; n.Type = sbom.Node_PACKAGE }, cdx.ComponentTypePlatform},
```
Comment (puerco) [inline pkg/native/serializers/serializer_cdx_test.go:33]: These new types are from CDX 1.5, they should be added to the proto enumeration too.

**Review by puerco — COMMENTED** [review]
Thanks for working on this @veramine!

I am in favor of getting this in and, especially, getting it in before 0.2.0. The PR is looking great, I added some initial comments below regarding:

1. The field should be a list as in SPDX it has a cardinality of more than one (note that this is a bug coming from the original onesbom proto).
2. We should define in the proto enum the new component types from CDX 1.5
3. Finally, remember to keep you commits as logical units of change that we can roll back if we need to. Especially please squash the last 7 or so dealing with whitespace changes.

This is great, thank you!

**Comment by veramine** [comment]
Cool thanks I'll be working on this today.  Here are the three lists:

SPDX 2.3 "Purpose" field values:  APPLICATION | FRAMEWORK | LIBRARY | CONTAINER | OPERATING-SYSTEM | DEVICE | FIRMWARE | SOURCE | ARCHIVE | FILE | INSTALL | OTHER

CycloneDX 1.5 "Component types" field values:  "application", "framework", "library", "container", "platform", "operating-system", "device", "device-driver", "firmware", "file", "machine-learning-model", "data"

SPDX 3.0 "SoftwarePurpose" field values: application, archive, bom, configuration, container, data, device, documentation, evidence, executable, file, firmware, framework, install, library, manifest, model, module, operatingSystem, other, patch, requirement, source, specification, test

File: api/sbom.proto (hunk 1 above)
Comment (veramine) [reply api/sbom.proto:58]: cool, no worries.  glad we have a chance to get this right.  before 0.2.0 is the time!

**Comment by veramine** [comment]
Closing this in favor of https://github.com/bom-squad/protobom/pull/136.

**Pull request closed without merging**